### PR TITLE
Add default book title and authors

### LINKED ISSUE #4912: Books in book.txt do not work
### Type of bug

Other unexpected behaviour

### `/ess dump all` output

https://essentialsx.net/dump.html?id=377206bc9a9649cd8d67b0685541bdf8

### Error log (if applicable)

_No response_

### Bug description

When giving myself a book defined in book.txt, it says "invalid book tag".

Book.txt:
`#starter`
`This is the starter book`

Command: 
`/i writtenbook 1 book:starter`

Result:
image

However, if I give myself a writable book instead, the text shows (typing /book switches it back and forth between a written and writable book, and the text switches back and forth between the intended text and "invalid book tag").
image

Thanks!

### Steps to reproduce

1. Define a book in book.txt
2. Use the command `/give <player> writtenbook 1 book:<bookname>`
3. Observe the "invalid book tag" error

### Expected behaviour

The text will show

### Actual behaviour

The "invalid book tag" error shows

### PR DESCRIPTION
**Pull request by SavageAvocado**
**Information**

This PR fixes https://github.com/EssentialsX/Essentials/issues/4912.

**Details**

**Proposed fix:**
Fixes "invalid book tag" issue by setting a default author & title for written books.

**Environments tested:**

OS: Windows 10

Java version: 17.0.1

- [X] Most recent Paper version (1.18.2)
- [ ] CraftBukkit/Spigot/Paper 1.12.2
- [ ] CraftBukkit 1.8.8

**Demonstration:**

Before (Tested with EssentialsX 2.20.0)
https://youtu.be/D7pgr_nSkAI

After (Tested by using the compiled version with this commit)
https://youtu.be/MS0klLUz2A0

### CODE REVIEW

File: Essentials/src/main/java/com/earth2me/essentials/MetaItemStack.java [hunk 1]
```diff
@@ -10,10 +10,7 @@
 import com.earth2me.essentials.utils.VersionUtil;
 import com.google.common.base.Joiner;
 import net.ess3.api.IEssentials;
-import org.bukkit.Color;
-import org.bukkit.DyeColor;
-import org.bukkit.FireworkEffect;
-import org.bukkit.Material;
+import org.bukkit.*;
```
Comment (pop4959) [inline Essentials/src/main/java/com/earth2me/essentials/MetaItemStack.java]: Don't use wildcard imports. You can configure your IDE not to do this.

File: Essentials/src/main/java/com/earth2me/essentials/MetaItemStack.java [hunk 2]
```diff
@@ -45,6 +42,8 @@
 import static com.earth2me.essentials.I18n.tl;
 
 public class MetaItemStack {
+    private static final String DEFAULT_BOOK_AUTHOR = Bukkit.getConsoleSender().getName();
```
Comment (pop4959) [inline Essentials/src/main/java/com/earth2me/essentials/MetaItemStack.java]: Not sure I like this... definitely more of an opinionated thing but I'd much rather it default to the player requesting/spawning the book, rather than console. Either that or make this configurable.

File: Essentials/src/main/java/com/earth2me/essentials/MetaItemStack.java [hunk 3]
```diff
@@ -216,11 +215,20 @@ public void addStringMeta(final CommandSource sender, final boolean allowUnsafe,
...
-        } else if (split.length > 1 && split[0].equalsIgnoreCase("book") && MaterialUtil.isEditableBook(stack.getType()) && (hasMetaPermission(sender, "book",true, true, ess) || hasMetaPermission(sender, "chapter-" + split[1].toLowerCase(Locale.ENGLISH), true, true, ess))) {
+        } else if (split.length > 1 && split[0].equalsIgnoreCase("book") && MaterialUtil.isEditableBook(stack.getType()) && (hasMetaPermission(sender, "book", true, true, ess) || hasMetaPermission(sender, "chapter-" + split[1].toLowerCase(Locale.ENGLISH), true, true, ess))) {
             final BookMeta meta = (BookMeta) stack.getItemMeta();
             final IText input = new BookInput("book", true, ess);
             final BookPager pager = new BookPager(input);
 
+            // This fix only applies to written books.
+            if (stack.getType() == WRITTEN_BOOK) {
+                // Written books require an author and a title. https://bugs.mojang.com/browse/MC-59153
+                meta.setAuthor(DEFAULT_BOOK_AUTHOR);
```
Comment (pop4959) [inline Essentials/src/main/java/com/earth2me/essentials/MetaItemStack.java]: We should probably try to use author meta here if applicable, and only default when it is not available. See the Essentials Item Meta format for books [here](https://wiki.mc-ess.net/wiki/Item_Meta#Books) on the old wiki.

**Review by pop4959 — DISMISSED** [review]
Some minor things to touch up on, but this is looking great!

File: Essentials/src/main/java/com/earth2me/essentials/MetaItemStack.java [hunk 4]
```diff
@@ -216,11 +217,22 @@ public void addStringMeta(final CommandSource sender, final boolean allowUnsafe,
...
+        } else if (split.length > 1 && split[0].equalsIgnoreCase("book") && MaterialUtil.isEditableBook(stack.getType()) && (hasMetaPermission(sender, "book", true, true, ess) || hasMetaPermission(sender, "chapter-" + split[1].toLowerCase(Locale.ENGLISH), true, true, ess))) {
             final BookMeta meta = (BookMeta) stack.getItemMeta();
             final IText input = new BookInput("book", true, ess);
             final BookPager pager = new BookPager(input);
-
+            // This fix only applies to written books.
+            if (stack.getType() == WRITTEN_BOOK) {
+                // Written books require an author and a title. https://bugs.mojang.com/browse/MC-59153
+                if (!meta.hasAuthor()) {
+                    meta.setAuthor(sender.getDisplayName());
```
Comment (pop4959) [inline Essentials/src/main/java/com/earth2me/essentials/MetaItemStack.java]: Particular reason for this being display name rather than username?

File: Essentials/src/main/java/com/earth2me/essentials/MetaItemStack.java [hunk 5]
```diff
@@ -216,11 +217,22 @@ public void addStringMeta(final CommandSource sender, final boolean allowUnsafe,
...
+                }
+                if (!meta.hasTitle()) {
+                    final String title = StringUtils.capitalize(split[1].toLowerCase(Locale.ROOT));
```
Comment (pop4959) [inline Essentials/src/main/java/com/earth2me/essentials/MetaItemStack.java]: Should this not format the same way as for title meta below?

Comment (mdcfe) [reply Essentials/src/main/java/com/earth2me/essentials/MetaItemStack.java]: I think the formatting is fine - `book:` values are often written in lowercase for convenience.

This will need updating to use our `StringUtil` as we no longer depend on Aapche Commons.

File: Essentials/src/main/java/com/earth2me/essentials/MetaItemStack.java (hunk 4 above)
Comment (mdcfe) [reply Essentials/src/main/java/com/earth2me/essentials/MetaItemStack.java]: I would also prefer we used the username here.

**Review by mdcfe — DISMISSED** [review]

File: Essentials/src/main/java/com/earth2me/essentials/MetaItemStack.java [hunk 6]
```diff
@@ -216,11 +217,21 @@ public void addStringMeta(final CommandSource sender, final boolean allowUnsafe,
...
+        } else if (split.length > 1 && split[0].equalsIgnoreCase("book") && MaterialUtil.isEditableBook(stack.getType()) && (hasMetaPermission(sender, "book", true, true, ess) || hasMetaPermission(sender, "chapter-" + split[1].toLowerCase(Locale.ENGLISH), true, true, ess))) {
             final BookMeta meta = (BookMeta) stack.getItemMeta();
             final IText input = new BookInput("book", true, ess);
             final BookPager pager = new BookPager(input);
-
+            // This fix only applies to written books - which require an author and a title. https://bugs.mojang.com/browse/MC-59153
+            if (stack.getType() == WRITTEN_BOOK) {
+                if (!meta.hasAuthor()) {
+                    // The sender can be null when this method is called from {@link  com.earth2me.essentials.signs.EssentialsSign#getItemMeta(ItemStack, String, IEssentials)}
+                    meta.setAuthor(sender == null ? Bukkit.getConsoleSender().getName() : sender.getPlayer().getName());
```
Comment (mdcfe) [inline Essentials/src/main/java/com/earth2me/essentials/MetaItemStack.java]: I'm not sure if we want to use the console sender name, or if it would be better to have a translation key like `bookDefaultAuthor` that servers/translators can override? cc @pop4959 @JRoy

File: Essentials/src/main/java/com/earth2me/essentials/MetaItemStack.java [hunk 7]
```diff
@@ -286,9 +297,9 @@ public void addStringMeta(final CommandSource sender, final boolean allowUnsafe,
                 final String input = color[0];
                 if (input.startsWith("#")) { // Hex
                     meta.setColor(Color.fromRGB(
-                        Integer.valueOf(input.substring(1, 3), 16),
-                        Integer.valueOf(input.substring(3, 5), 16),
-                        Integer.valueOf(input.substring(5, 7), 16)));
+                            Integer.valueOf(input.substring(1, 3), 16),
+                            Integer.valueOf(input.substring(3, 5), 16),
+                            Integer.valueOf(input.substring(5, 7), 16)));
```
Comment (mdcfe) [inline Essentials/src/main/java/com/earth2me/essentials/MetaItemStack.java]: Undo this unnecessary diff

File: Essentials/src/main/java/com/earth2me/essentials/MetaItemStack.java (hunk 6 above)
Comment (JRoy) [reply Essentials/src/main/java/com/earth2me/essentials/MetaItemStack.java]: Agree with the translation

Comment (pop4959) [reply Essentials/src/main/java/com/earth2me/essentials/MetaItemStack.java]: I see no reason not to, I think in all cases where a non-player is spawning books it's likely they would want to set the `author` meta anyways, but this seems good to do anyway.

File: Essentials/src/main/java/com/earth2me/essentials/MetaItemStack.java (hunk 7 above)
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/MetaItemStack.java]:
```suggestion
                        Integer.valueOf(input.substring(1, 3), 16),
                        Integer.valueOf(input.substring(3, 5), 16),
                        Integer.valueOf(input.substring(5, 7), 16)));
```

File: Essentials/src/main/java/com/earth2me/essentials/signs/EssentialsSign.java [hunk 8]
```diff
@@ -46,7 +46,7 @@ protected static boolean checkIfBlockBreaksSigns(final Block block) {
         if (MaterialUtil.isSignPost(sign.getType()) && isValidSign(new BlockSign(sign))) {
             return true;
         }
-        final BlockFace[] directions = new BlockFace[] {BlockFace.NORTH, BlockFace.EAST, BlockFace.SOUTH, BlockFace.WEST};
+        final BlockFace[] directions = new BlockFace[]{BlockFace.NORTH, BlockFace.EAST, BlockFace.SOUTH, BlockFace.WEST};
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/signs/EssentialsSign.java]:
```suggestion
        final BlockFace[] directions = new BlockFace[] {BlockFace.NORTH, BlockFace.EAST, BlockFace.SOUTH, BlockFace.WEST};
```

**Review by JRoy — APPROVED** [review]
